OSLObject : Support <attr:> substitutions

danieldresser-ie opened the pull request:
Not sure I'm really a fan of how this turned out ... feels a bit like we're redoing hashing work more than we should be, and there's a couple TODOs I've left on troublesome bits.  But it does seem to be working, and probably more meaningful to talk about it with a working prototype ... maybe this approach is good enough.

johnhaddon commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -84,9 +104,20 @@ struct ShadingEngineCacheGetterKey
+			
+			ConstCompoundObjectPtr attributes = shader->attributes();
+			const ShaderNetwork *network = findOSLShaderNetwork( attributes.get() );
+			if( network && network->size() )
```
Why do we care about size? Won't `ShaderNetwork` deal with that?

danieldresser-ie replied:
Fixed in 3c3d696

johnhaddon commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -84,9 +104,20 @@ struct ShadingEngineCacheGetterKey
+			const ShaderNetwork *network = findOSLShaderNetwork( attributes.get() );
+			if( network && network->size() )
+			{
+				network->hashSubstitutions( targetAttributes, hash );
```
Since we're going out of our way to "deduplicate" ShadingEngines as much as possible here, it'd be good to have some test coverage for that.

danieldresser-ie replied:
Is this the sort of thing you were thinking of? e3df5be

johnhaddon replied:
Yep, exactly!

johnhaddon commented on src/GafferOSL/ShadingEngine.cpp:
```diff
@@ -932,11 +932,16 @@ static const T *varyingValue( const IECore::CompoundData *points, const char *na
 {
 	ShaderNetworkPtr networkCopy = shaderNetwork->copy();
 	IECoreScene::ShaderNetworkAlgo::convertOSLComponentConnections( networkCopy.get(), OSL_VERSION );
+	if( targetAttributes )
+	{
+		networkCopy->applySubstitutions( targetAttributes );
+	}
```
I'd assumed we'd be doing this in the cache getter in OSLShader.cpp, since we were trying to keep the substitution stuff outside of ShadingEngine.

danieldresser-ie replied:
Fixed in 6b0990f ( using r-value reference constructor )

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -410,7 +423,10 @@ void OSLObject::compute( Gaffer::ValuePlug *output, const Gaffer::Context *conte
 		}
 
-		ConstShadingEnginePtr shadingEngine = this->shadingEngine( context );
+		// TODO - think I probably do need to pass the attributes in here, which means also sticking the
```
I guess this is for the case where a substitution affects `ShadingEngine::needsAttribute()`? That seems quite unlikely, but not impossible of course. I suppose if we did the `lockgeom=false` thing for parameters containing substitutions, then that would take care of this without needing to pass the attributes at this point?

danieldresser-ie replied:
Yeah, I think lockgeom=false should take care of this.

johnhaddon replied:
I thought one of the reasons we'd come back to this approach where we substitute the network upfront was to avoid the lack of constant folding implied by `lockgeom=false`?

johnhaddon replied:
If we do need to pass the attributes to `shadingEngine()` then perhaps it makes sense for `OSLObject::ShadingEngine()` to get the attributes itself internally (providing them as an optional output variable)? Since we want to use substitutions everywhere except `adjustBounds()`, it seems like it might be less error-prone for `shadingEngine()` to take full responsibility?

danieldresser-ie replied:
> I thought one of the reasons we'd come back to this approach where we substitute the network upfront was to avoid the lack of constant folding implied by lockgeom=false?

I think we want to make it an exception to evaluate a ShaderNetwork with unsubstituted substitutions, but we're not evaluating here, just using needsAttribute, so this case is basically identical to adjustBounds

The proposal I gave on adjustBounds applies here too, so if using the STRING_PARAMETER_VALUE_UNKNOWN token to label parameters that shouldn't be locked is acceptable, then we don't need to pass the attributes to shadingEngine.

johnhaddon reviewed (commented):
> Not sure I'm really a fan of how this turned out ... feels a bit like we're redoing hashing work more than we should be

Seeing this does make me more interested in the idea of doing the substitutions in `ShadingEngine::shade()`. Maybe the `lockgeom=false` thing is worth a quick experiment?

danieldresser-ie commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -310,9 +317,10 @@ void OSLObject::hashProcessedObject( const ScenePath &path, const Gaffer::Contex
 	{
 		h.append( inPlug()->fullTransformHash( path ) );
 	}
-	if( useAttributesPlug()->getValue() )
+
+	if( useAttributes )
 	{
-		h.append( inPlug()->fullAttributesHash( path ) );
+		gafferAttributes->hash( h );
```
We can actually do better ... I'm pretty sure we can just omit this?  We're already including `shadingEngine->hash`, which includes all the parameters that we have baked substitutions into already.

danieldresser-ie replied:
Oh, I was just thinking about substitutions, and not thinking about how this was already needed for regular getattributes.  I have now fixed this in cab7548

In the course of implementing this, I compared it to what it was previously using, CompoundObject::hash ... looking at that was ... rather alarming?  I haven't measured yet how often Gaffer is actually using this, but I really wouldn't expect this hash to allocate a vector of iterators, and do a pile of string comparisons, just for the sake of inter-process consistency which is unnecessary for Gaffer.

johnhaddon replied:
> I compared it to what it was previously using, CompoundObject::hash

It wasn't using `CompoundObject::hash()` before - it was using `fullAttributesHash()`, which is a `ValuePlug::hash()` rather than an `Object::hash()`.

danieldresser-ie commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -382,7 +391,11 @@ bool OSLObject::adjustBounds() const
 	}
 
-	ConstShadingEnginePtr s = shadingEngine( Context::current() );
+	// TODO - this is technically wrong, since in theory I think you could build a shader where all
+	// writes to P get constant folded out when the attribute substutitions aren't applied, but
+	// does actually write to P when the attributes are substituted.  Since adjustBounds is called
+	// in a global scope however, it's hard to do this properly.
```
OK, if we really want to do this right:

ShadingEngine declares a special token STRING_PARAMETER_VALUE_UNKNOWN.  Evaluating shade() on a ShadingEngine containing these tokens throws an exception, but calls to hasDeformation yield correct results ( implemented using lockgeom=false ).

The getter for ShadingEngines inside OSLShader can then be modified to replace strings that require substitutions with this token when no attributes are provided ... I guess this might require a special sentinel value for the attributes that OSLObject::adjustBounds passes to this getter, to indicate "the attribute values are unknown" vs "the attribute values aren't provided" ( ie. useAttributes is set false ).

This approach seems like it's kinda complicated, but would provide correct behaviour.

johnhaddon replied:
That does sound like it would work. But I'm willing to bet that this is enough of a corner case that it's not worth the effort or complexity at this point. My vote would be to just add a short comment linking to this discussion.

danieldresser-ie commented on include/GafferOSL/OSLShader.h:
```diff
@@ -66,7 +66,7 @@ class GAFFEROSL_API OSLShader : public GafferScene::Shader
 		void reloadShader() override;
 
-		ConstShadingEnginePtr shadingEngine() const;
+		ConstShadingEnginePtr shadingEngine( const IECore::CompoundObject *targetAttributes = nullptr ) const;
```
Fixed in e949fc8

danieldresser-ie commented on python/GafferOSLTest/OSLObjectTest.py:
```diff
@@ -993,6 +993,41 @@ def testAttributes( self ):
+
+		o = GafferOSL.OSLObject()
+		o["in"].setInput( a["out"] )
+		o["useAttributes"].setValue( True )
```
Fixed in d3a9037

danieldresser-ie commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -293,7 +293,14 @@ bool OSLObject::affectsProcessedObject( const Gaffer::Plug *input ) const
 void OSLObject::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
 {
-	ConstShadingEnginePtr shadingEngine = this->shadingEngine( context );
+	bool useAttributes = useAttributesPlug()->getValue();
```
Fixed in 3c3d696

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -310,9 +316,23 @@ void OSLObject::hashProcessedObject( const ScenePath &path, const Gaffer::Contex
+			{
+				if ( !i.second )
+				{
+					throw Exception( "Cannot compute hash from a CompoundObject will NULL data pointers!" );
```
There's a typo in the message here, but I don't think we need this check anyway - everywhere else just assumes attributes have to be non-null.

danieldresser-ie replied:
Oh, huh, I didn't want to change behaviour more than necessary, so I just copied this from Cortex ... I guess I should probably fix the typo here too: https://github.com/ImageEngine/cortex/blob/main/src/IECore/CompoundObject.cpp#L77

danieldresser-ie replied:
Oh, doh.  Just saw the later comment about how it was never calling CompoundObject::hash in the first place.   Sorry for the confusion, though I suppose maybe I should still fix the Cortex typo

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -310,9 +316,23 @@ void OSLObject::hashProcessedObject( const ScenePath &path, const Gaffer::Contex
+	if( gafferAttributes )
 	{
-		h.append( inPlug()->fullAttributesHash( path ) );
+		for( const auto &i : gafferAttributes->members() )
```
I think we should adopt structured bindings as our standard idiom for this sort of loop, instead of using `i.first` and `i.second` :

```
for( const auto &[name, value] : gafferAttributes->members() )
{
    ...
```

It makes the rest of the loop a fair bit more readable...

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -310,9 +316,23 @@ void OSLObject::hashProcessedObject( const ScenePath &path, const Gaffer::Contex
+				{
+					throw Exception( "Cannot compute hash from a CompoundObject will NULL data pointers!" );
+				}
+				// \todo - why are we hashing the value instead of the address of the InternedString?
+				// More importantly, why does the default hash of CompoundObject do so much work?
+				h.append( i.first.value() );
+				i.second->hash( h );
```
The comment about `CompoundObject::hash()` is really out of context here, where we're not even calling it - I don't think it will make much sense in future. I'd have to do a bit of a review to see whether or not we hash values instead of addresses in Gaffer - there are arguments both ways (with the argument for stable hashes being easier debugging).

johnhaddon commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -84,9 +104,19 @@ struct ShadingEngineCacheGetterKey
 
-	ShadingEngineCacheGetterKey( const OSLShader *s )
-		:	shader( s ), hash( s->attributesHash() )
+	ShadingEngineCacheGetterKey( const OSLShader *s, const CompoundObject *targetA )
```
The `targetA` name seems to be a hangover from previous naming - should just be `substitutions`.

johnhaddon reviewed (commented):
I've been over the latest changes and made a few comments - looks maybe ithe approach to `compute( resampledNamesPlug )` is the main outstanding thing...

danieldresser-ie commented:
OK, I think this about ready to go.  I've addressed the last round of feedback ( taking the direction we discussed this morning ), and rebased onto 1.1_maintenance, and added a Changes entry.

There one new commit of fixes that will need squashing after review, but other than that should be good to go.
